node/service/src/relay_chain_selection.rs

```diff
@@ -338,16 +338,41 @@ impl<B> SelectChain<PolkadotBlock> for SelectRelayChain<B>
 				.map_err(|e| ConsensusError::Other(Box::new(e)))?
 			{
 				// No approved ancestors means target hash is maximal vote.
-				None => (target_hash, target_number),
-				Some((s_h, s_n)) => (s_h, s_n),
+				None => (target_hash, target_number, Vec::new()),
+				Some(HighestApprovedAncestorBlock {
+					number, hash, descriptions
+				}) => (hash, number, descriptions),
 			}
 		};
 
-		let lag = initial_leaf_number.saturating_sub(subchain_number);
-		self.metrics.note_approval_checking_finality_lag(lag);
+		// Prevent sending flawed data to the controller.
+		if Some(subchain_block_descriptions.len() as _) != subchain_number.checked_sub(target_number) {
```

Comment:
1. what is `the controller`?
2. isn't that a logical bug? why not a `debug_assert!`? looking at the code that does `candidates.clear() if !entry.fully_approved()`, it's not clear to me why this invariant holds, perhaps a test would help clarify this ;)

Reply:
`controller` -> `coordinator`